http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/internal/http2/Http2FrameExceptionHandler.java

```diff
@@ -0,0 +1,100 @@
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.http.nio.netty.internal.http2;
+
+import static software.amazon.awssdk.http.nio.netty.internal.ChannelAttributeKey.CHANNEL_POOL_RECORD;
+import static software.amazon.awssdk.http.nio.netty.internal.ChannelAttributeKey.PING_STATUS;
+import static software.amazon.awssdk.http.nio.netty.internal.NettyConfiguration.HTTP2_CONNECTION_PING_TIMEOUT_SECONDS;
+
+import io.netty.channel.Channel;
+import io.netty.channel.ChannelHandler;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelInboundHandlerAdapter;
+import io.netty.handler.codec.http2.DefaultHttp2PingFrame;
+import io.netty.handler.codec.http2.Http2PingFrame;
+import io.netty.handler.timeout.ReadTimeoutException;
+import io.netty.handler.timeout.WriteTimeoutException;
+import java.io.IOException;
+import java.util.concurrent.TimeUnit;
+import software.amazon.awssdk.annotations.SdkInternalApi;
+import software.amazon.awssdk.utils.Logger;
+
+@ChannelHandler.Sharable
+@SdkInternalApi
+public final class Http2FrameExceptionHandler extends ChannelInboundHandlerAdapter {
+    private static final Http2FrameExceptionHandler INSTANCE = new Http2FrameExceptionHandler();
+    private static final Http2PingFrame DEFAULT_PING_FRAME = new DefaultHttp2PingFrame(0);
+    private static final Logger log = Logger.loggerFor(Http2FrameExceptionHandler.class);
+
+    private Http2FrameExceptionHandler() {
+    }
+
+    public static Http2FrameExceptionHandler create() {
+        return INSTANCE;
+    }
+
+    @Override
+    public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) {
+        if ((cause instanceof ReadTimeoutException || cause instanceof WriteTimeoutException) && ctx.channel().parent() != null) {
+            healthCheckIfNotAlready(ctx);
+        }
+
+        ctx.fireExceptionCaught(cause);
+    }
+
+    private void healthCheckIfNotAlready(ChannelHandlerContext ctx) {
+        Channel parent = ctx.channel().parent();
+
+        // No need to do another health check if there is a PING in flight
+        if (PingStatus.IN_FLIGHT.equals(parent.attr(PING_STATUS).get())) {
+            return;
+        }
+
+        log.debug(() -> "Read/WriteTimeoutException occurred, sending PING frame to " + parent);
+        parent.attr(PING_STATUS).set(PingStatus.IN_FLIGHT);
+
+        parent.writeAndFlush(DEFAULT_PING_FRAME).addListener(res -> {
+            if (!res.isSuccess()) {
+                log.debug(() -> "Failed to write and flush PING frame to connection", res.cause());
+                closeH2Connection(parent);
+            }
+        });
+        parent.eventLoop().schedule(() -> timeoutHealthCheck(parent), HTTP2_CONNECTION_PING_TIMEOUT_SECONDS,
+                                TimeUnit.SECONDS);
```

Comment:
Should this only be done when the send-ping succeeds?

Reply:
It was actually done within the listener in the previous revision and I moved it out because write itself might take a long time as well. 

I tried adding a Write timeout handler there and it did not work somehow.